File: 06_Searching/src/Minimum_Absolute_Sum_Difference_1818.js
const minAbsoluteSumDiff = function(nums1, nums2) {
    let mod = 1000000007;
    let totalSum = 0;
    let maxImprovement = 0;

    let copyArr = [...nums1].sort((a, b) => a - b);

    for (let i = 0; i < nums1.length; i++) {
        let cDiff = Math.abs(nums1[i] - nums2[i]);
        totalSum += cDiff;
        maxImprovement = 0;

        let nearby = find (copyArr,nums2[i]);

        if (nearby < nums2.length) {
            let nDiff = Math.abs(copyArr[nearby] - nums2[i]);
            maxImprovement = Math.max(maxImprovement, (cDiff - nDiff));
        }

        if (nearby > 0) {
            let nDiff = Math.abs(copyArr[nearby-1] - nums2[i]);
            maxImprovement = Math.max(maxImprovement, (cDiff - nDiff));
        }
    }

    function find (arr,target) {
        let start = 0;
        let end = arr.length-1;

        while (start < end) {
            let mid = Math.floor(start + (end-start)/2);

            if (arr[mid] === target) {
                return mid;
            } else if (arr[mid] > target) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }

        return end;
    }

};

File: 06_Searching/src/Minimum_Absolute_Sum_Difference_1818.java
import java.util.Arrays;

public class Minimum_Absolute_Sum_Difference_1818 {
    public static int minAbsoluteSumDiff(int[] nums1, int[] nums2) {
        int[] initialAbsoluteDiff = new int[nums1.length];
        int initialTotalSum = 0;

        for (int i = 0; i < initialAbsoluteDiff.length; i++) {
            initialAbsoluteDiff[i] = Math.abs(nums1[i] - nums2[i]);
            initialTotalSum += initialAbsoluteDiff[i];
        }

        Arrays.sort(nums1);

        int[] lowestDiff = new int[nums2.length];
        int maxDiff = 0;

        for (int i = 0; i < nums2.length; i++) {
            int temp = lowerBound(nums2[i], nums1);
            int tempDiff = Math.abs(nums2[i] - temp);

            if (tempDiff < initialAbsoluteDiff[i]) {
                maxDiff = Math.abs(tempDiff-initialAbsoluteDiff[i]);
            }
        }

        return initialTotalSum - maxDiff;
    }

    public static int lowerBound (int target, int[] arr) {
        int start = 0;
        int end = arr.length-1;

        while (start < end) {
            int mid = start + (end - start) / 2;

            if (arr[mid] == target) {
                return arr[mid];
            } else if (arr[mid] > target) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }

        int ans = Math.min()

        return arr[end];
    }

    public static void main(String[] args) {
        System.out.println(minAbsoluteSumDiff(new int[] {1,7,5}, new int[] {2,3,5}));
        System.out.println(minAbsoluteSumDiff(new int[] {2,4,6,8,10}, new int[] {2,4,6,8,10}));
        System.out.println(minAbsoluteSumDiff(new int[] {1,10,4,4,2,7}, new int[] {9,3,5,1,7,4}));
    }
}
